### PR TITLE
Task-50874: Implement add target backend

### PR DESCRIPTION
**Pull request by hassenelkamel-exo**

### CODE REVIEW

File: services/src/main/java/org/exoplatform/news/service/impl/NewsTargetingServiceImpl.java [hunk 1]
```diff
@@ -135,11 +137,26 @@ public void deleteNewsTargets(String newsId, String currentUserId) throws Illega
     deleteNewsTargets(newsId);
   }
 
+  @Override
+  public Metadata createMetadata(Metadata metadata, long userIdentityId) throws IllegalArgumentException, IllegalAccessException {
+    if (!NewsUtils.canCreateTarget(identityManager, userIdentityId)) {
+      throw new IllegalAccessException("User " + userIdentityId + " not authorized to delete news targets");
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/impl/NewsTargetingServiceImpl.java]:
```suggestion
      throw new IllegalAccessException("User " + userIdentityId + " not authorized to add news target");
```

File: webapp/src/main/webapp/news-publish-targets-management/components/NewsPublishTargetsManagementDrawer.vue [hunk 2]
```diff
@@ -1,5 +1,5 @@
 <!--
-Copyright (C) 2022 eXo Platform SAS.
+Copyright (C) 2021 eXo Platform SAS.
```
Comment (azayati) [inline webapp/src/main/webapp/news-publish-targets-management/components/NewsPublishTargetsManagementDrawer.vue]: should be 2022

File: webapp/src/main/webapp/news-publish-targets-management/components/NewsPublishTargetsManagementDrawer.vue [hunk 3]
```diff
@@ -97,14 +107,68 @@ export default {
         return '';
```
Comment (azayati) [inline webapp/src/main/webapp/news-publish-targets-management/components/NewsPublishTargetsManagementDrawer.vue]: transform to return this.targetName && !this.targetName.trim().match(/^[\w\-\s]+$/) && this.targetName.length > 0 ? this.$t('news.list.settings.name.errorMessage') : ''

File: webapp/src/main/webapp/news-publish-targets-management/components/NewsPublishTargetsManagementDrawer.vue [hunk 4]
```diff
@@ -97,14 +107,68 @@ export default {
...
       }
     },
+    disabled() {
+      return this.checkAlphanumeric !== '' || (this.targetName && this.targetName.length === 0) || (this.targetName && this.targetName.length >0 && this.sameTargetError);
```
Comment (azayati) [inline webapp/src/main/webapp/news-publish-targets-management/components/NewsPublishTargetsManagementDrawer.vue]:
```suggestion
      return this.checkAlphanumeric ! == '' || this.targetName.length === 0 || this.sameTargetError;
```

File: webapp/src/main/webapp/news-publish-targets-management/components/NewsPublishTargetsManagementDrawer.vue [hunk 5]
```diff
@@ -97,14 +107,68 @@ export default {
...
+      } else {
+        this.$refs.newsPublishTargetsManagementDrawer.endLoading();
+      }
+    },
+    targetName(newVal, oldVal) {
+      if (newVal && newVal.length > 0 && oldVal && newVal === oldVal) {
+        this.sameTargetError =  true;
+      } else {
+        this.sameTargetError =false;
+      }
```
Comment (azayati) [inline webapp/src/main/webapp/news-publish-targets-management/components/NewsPublishTargetsManagementDrawer.vue]:
```suggestion
      this.sameTargetError =  newVal && newVal.length > 0 && oldVal && newVal === oldVal;
   
```

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 6]
```diff
@@ -202,4 +210,38 @@ public Response undoDeleteTarget(@Context HttpServletRequest request,
...
+                  @ApiResponse(code = HTTPStatus.BAD_REQUEST, message = "Invalid query input"),
+                  @ApiResponse(code = HTTPStatus.FORBIDDEN, message = "Forbidden operation"),
+                  @ApiResponse(code = HTTPStatus.UNAUTHORIZED, message = "User not authorized to create news target"),
+                  @ApiResponse(code = HTTPStatus.CONFLICT, message = "Conflict operation") }
+  )
+  public Response createNewsTarget(@Context HttpServletRequest request,
+                                       @ApiParam(value = "News target to create", required = true) Metadata newsTarget) {
+    long userIdentityId = NewsUtils.getCurrentUserIdentityId(identityManager);
+    try {
+      Metadata addedNewsTarget = newsTargetingService.createMetadata(newsTarget, userIdentityId);
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java]: I think that you should juste send NewsTargetingEntity as a parameter, then instanciate the corresponding metadata to be sent to metadata Service

Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java]:
```suggestion
      Metadata addedNewsTarget = newsTargetingService.createNewsTartget(newsTarget, userIdentityId);
```

File: services/src/main/java/org/exoplatform/news/service/NewsTargetingService.java [hunk 7]
```diff
@@ -101,4 +102,14 @@
...
 
+  /**
+   * Create a news Metadata
+   * 
+   * @param metadata {@link Metadata}
+   * @param userIdentityId {@link Identity} identifier of the creator
+   * @return created {@link Metadata}
+   * @throws IllegalArgumentException when user create a metadata that already exist
+   */
+  Metadata createMetadata(Metadata metadata, long userIdentityId) throws IllegalArgumentException, IllegalAccessException ;
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/NewsTargetingService.java]:
```suggestion
  Metadata createNewsTarget(NewsTargetingEntity newsTargetingEntity) throws ... ;
```

File: services/src/main/java/org/exoplatform/news/utils/NewsUtils.java [hunk 8]
```diff
@@ -123,6 +124,15 @@ public static boolean canPublishNews(org.exoplatform.services.security.Identity
   public static boolean canDeleteTargetNews(org.exoplatform.services.security.Identity currentIdentity) {
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/utils/NewsUtils.java]:
```suggestion
  public static boolean canManageNewsPublishTargets(org.exoplatform.services.security.Identity currentIdentity) {
```
to be used for all manage operations authorizations

**Review by azayati — CHANGES_REQUESTED** [review]
see review

File: services/src/main/java/org/exoplatform/news/service/NewsTargetingService.java [hunk 9]
```diff
@@ -101,4 +102,14 @@
...
 
+  /**
+   * Create a news Metadata
+   * 
+   * @param newsTargetingEntity {@link NewsTargetingEntity}
+   * @param currentIdentity {@link Identity} technical identifier
+   * @return created {@link Metadata}
+   * @throws IllegalArgumentException when user create a metadata that already exist
+   */
+  Metadata createNewsTarget(NewsTargetingEntity newsTargetingEntity, org.exoplatform.services.security.Identity currentIdentity) throws IllegalArgumentException, IllegalAccessException ;
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/NewsTargetingService.java]:
```suggestion
  Metadata createNewsTarget(NewsTargetingEntity newsTargetingEntity, org.exoplatform.services.security.Identity currentIdentity) throws IllegalArgumentException, IllegalAccessException;
```

File: services/src/main/java/org/exoplatform/news/service/NewsTargetingService.java [hunk 10]
```diff
@@ -101,4 +102,14 @@
    */
   void deleteNewsTargets(String newsId);
 
+  /**
+   * Create a news Metadata
+   * 
+   * @param newsTargetingEntity {@link NewsTargetingEntity}
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/NewsTargetingService.java]:
```suggestion
   * @param newsTargetingEntity {@link News} targeting entity
```

File: services/src/main/java/org/exoplatform/news/service/NewsTargetingService.java [hunk 11]
```diff
@@ -101,4 +102,14 @@
...
+   * @param currentIdentity {@link Identity} technical identifier
+   * @return created {@link Metadata}
+   * @throws IllegalArgumentException when user create a metadata that already exist
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/NewsTargetingService.java:111]: add also IllegalAccessException

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 12]
```diff
@@ -202,4 +204,38 @@ public Response undoDeleteTarget(@Context HttpServletRequest request,
...
+                  @ApiResponse(code = HTTPStatus.CONFLICT, message = "Conflict operation") }
+  )
+  public Response createNewsTarget(@Context HttpServletRequest request,
+                                       @ApiParam(value = "News target to create", required = true) NewsTargetingEntity newsTargetingEntity) {
+    org.exoplatform.services.security.Identity currentIdentity = ConversationState.getCurrent().getIdentity();
+    try {
+      Metadata addedNewsTarget = newsTargetingService.createNewsTarget(newsTargetingEntity, currentIdentity);
+      return Response.ok(addedNewsTarget).build();
+    } catch (IllegalAccessException e) {
+      LOG.warn("User '{}' is not authorized to create a news target with name " + currentIdentity.getUserId(), e);
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java]:
```suggestion
      LOG.warn("User '{}' is not authorized to create a news target with name " newsTargetingEntity.getName(), currentIdentity.getUserId(), e);
```

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 13]
```diff
@@ -202,4 +204,38 @@ public Response undoDeleteTarget(@Context HttpServletRequest request,
...
+      return Response.status(Response.Status.UNAUTHORIZED).entity(e.getMessage()).build();
+    } catch (IllegalArgumentException e) {
+      LOG.warn("User '{}' can't create a news targets '{}' with the same name ", currentIdentity.getUserId(), newsTargetingEntity.getName(), e);
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java]:
```suggestion
      LOG.warn("User '{}' can't create a news targets '{}' with the same name " + newsTargetingEntity.getName(), currentIdentity.getUserId(), e);
```

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 14]
```diff
@@ -202,4 +204,38 @@ public Response undoDeleteTarget(@Context HttpServletRequest request,
...
+      return Response.status(Response.Status.CONFLICT).entity(e.getMessage()).build();
+    } catch (Exception e) {
+      LOG.error("Error creating a news target ", e);
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java:235]:
```suggestion
      LOG.error("Error when creating a news target ", e);
```

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 15]
```diff
@@ -202,4 +204,38 @@ public Response undoDeleteTarget(@Context HttpServletRequest request,
...
+      return Response.serverError().entity(e.getMessage()).build();
+    }
+  }
+
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java:239]:
```suggestion
```
Remove the empty line

File: services/src/main/java/org/exoplatform/news/service/NewsTargetingService.java [hunk 16]
```diff
@@ -101,4 +102,14 @@
    */
   void deleteNewsTargets(String newsId);
 
+  /**
+   * Create a news Metadata
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/NewsTargetingService.java:106]:
```suggestion
   * Create a news target
```

File: services/src/main/java/org/exoplatform/news/service/NewsTargetingService.java [hunk 17]
```diff
@@ -101,4 +102,14 @@
...
+   * 
+   * @param newsTargetingEntity {@link NewsTargetingEntity}
+   * @param currentIdentity {@link Identity} technical identifier
+   * @return created {@link Metadata}
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/NewsTargetingService.java:110]:
```suggestion
   * @return created {@link News} target {@link Metadata}
```

File: services/src/main/java/org/exoplatform/news/service/NewsTargetingService.java [hunk 18]
```diff
@@ -101,4 +102,14 @@
    */
   void deleteNewsTargets(String newsId);
 
+  /**
+   * Create a news Metadata
+   * 
+   * @param newsTargetingEntity {@link NewsTargetingEntity}
+   * @param currentIdentity {@link Identity} technical identifier
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/NewsTargetingService.java:109]:
```suggestion
   * @param currentIdentity current {@link Identity} attempting to create {@link News} target
```

File: services/src/main/java/org/exoplatform/news/service/impl/NewsTargetingServiceImpl.java [hunk 19]
```diff
@@ -135,13 +134,41 @@ public void deleteNewsTargets(String newsId, String currentUserId) throws Illega
...
   }
 
+  @Override
+  public Metadata createNewsTarget(NewsTargetingEntity newsTargetingEntity, org.exoplatform.services.security.Identity currentIdentity) throws IllegalArgumentException, IllegalAccessException {
+    Identity identity = identityManager.getOrCreateIdentity(OrganizationIdentityProvider.NAME, currentIdentity.getUserId());
+    long userIdentityId = identity == null ? 0 : Long.parseLong(identity.getId());
+    Metadata metadata = fromEntity(newsTargetingEntity);
+    metadata.setCreatorId(userIdentityId);
+    if (!NewsUtils.canManageNewsPublishTargets(currentIdentity)) {
+      throw new IllegalAccessException("User " + userIdentityId + " not authorized to add news targets");
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/impl/NewsTargetingServiceImpl.java]:
```suggestion
      throw new IllegalAccessException("User " + currentIdentity.getUserId() + " not authorized to create news target");
```

File: services/src/main/java/org/exoplatform/news/service/impl/NewsTargetingServiceImpl.java [hunk 20]
```diff
@@ -135,13 +134,41 @@ public void deleteNewsTargets(String newsId, String currentUserId) throws Illega
...
+    }
+    MetadataKey targetMetadataKey = new MetadataKey(METADATA_TYPE.getName(), metadata.getName(), 0);
+    Metadata storedMetadata = metadataService.getMetadataByKey(targetMetadataKey);
+    if (storedMetadata != null) {
+      throw new IllegalArgumentException("User " + userIdentityId + " not authorized to create news target with name "
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/impl/NewsTargetingServiceImpl.java]:
```suggestion
      throw new IllegalArgumentException("User " + currentIdentity.getUserId() + " not authorized to create news target with same name "
```

File: services/src/main/java/org/exoplatform/news/utils/NewsUtils.java [hunk 21]
```diff
@@ -120,7 +121,16 @@ public static boolean canPublishNews(org.exoplatform.services.security.Identity
     return currentIdentity != null && currentIdentity.isMemberOf(PLATFORM_WEB_CONTRIBUTORS_GROUP, PUBLISHER_MEMBERSHIP_NAME);
   }
 
-  public static boolean canDeleteTargetNews(org.exoplatform.services.security.Identity currentIdentity) {
+  public static boolean canManageNewsPublishTargets(org.exoplatform.services.security.Identity currentIdentity) {
+    return currentIdentity != null && currentIdentity.isMemberOf(PLATFORM_WEB_CONTRIBUTORS_GROUP, MANAGER_MEMBERSHIP_NAME);
+  }
+  
+  public static boolean canCreateTarget(IdentityManager identityManager, long userIdentityId) {
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/utils/NewsUtils.java]: useless to be removed

File: services/src/main/resources/locale/portlet/news/News_en.properties [hunk 22]
```diff
@@ -214,3 +214,4 @@ news.publishTargets.managementDrawer.btn.confirm=Confirm
 news.publishTargets.managementDrawer.btn.cancel=Cancel
 news.publishTargets.managementDrawer.placeholder.name=Enter the target name
 news.publishTargets.managementDrawer.placeholder.description=Enter a description here
+news.publishTargets.managementDrawer.sameTargetWarning=A target with the same name already exists.
```
Comment (azayati) [inline services/src/main/resources/locale/portlet/news/News_en.properties]:
```suggestion
news.publishTargets.managementDrawer.sameNewsTargetWarning=A news target with the same name already exists.
```

File: webapp/src/main/webapp/news-publish-targets-management/components/NewsPublishTargetsManagementDrawer.vue [hunk 23]
```diff
@@ -61,6 +61,12 @@ along with this program. If not, see <http://www.gnu.org/licenses/>.
               :max-length="targetDescriptionTextLength"
               class="targetDescription pt-0 " />
           </div>
+          <div v-if="sameTargetError" class="d-flex flex-row mt-4">
+            <v-icon class="warning--text">warning</v-icon>
+            <span class="ms-2 grey--text">
+              {{ $t('news.publishTargets.managementDrawer.sameTargetWarning') }}
```
Comment (azayati) [inline webapp/src/main/webapp/news-publish-targets-management/components/NewsPublishTargetsManagementDrawer.vue]:
```suggestion
              {{ $t('news.publishTargets.managementDrawer.sameNewsTargetWarning') }}
```

**Review by azayati — CHANGES_REQUESTED** [review]
see review

File: services/src/main/resources/locale/portlet/news/News_en.properties (hunk 22 above)
Comment (hassenelkamel-exo) [reply services/src/main/resources/locale/portlet/news/News_en.properties]: the message should be "A target with the same name already exists" like in US specification

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 24]
```diff
@@ -202,4 +203,38 @@ public Response undoDeleteTarget(@Context HttpServletRequest request,
     }
   }
 
+  @POST
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Produces(MediaType.APPLICATION_JSON)
+  @RolesAllowed("users")
+  @ApiOperation(
+          value = "Create a new target.",
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java:211]:
```suggestion
          value = "Create a news target",
```

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 25]
```diff
@@ -202,4 +203,38 @@ public Response undoDeleteTarget(@Context HttpServletRequest request,
...
+                  @ApiResponse(code = HTTPStatus.CONFLICT, message = "Conflict operation") }
+  )
+  public Response createNewsTarget(@Context HttpServletRequest request,
+                                       @ApiParam(value = "News target to create", required = true) NewsTargetingEntity newsTargetingEntity) {
+    org.exoplatform.services.security.Identity currentIdentity = ConversationState.getCurrent().getIdentity();
+    try {
+      Metadata addedNewsTarget = newsTargetingService.createNewsTarget(newsTargetingEntity, currentIdentity);
+      return Response.ok(addedNewsTarget).build();
+    } catch (IllegalAccessException e) {
+      LOG.warn("User '{}' is not authorized to create a news target with name ", newsTargetingEntity.getName(), currentIdentity.getUserId(), e);
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java:229]:
```suggestion
      LOG.warn("User '{}' is not authorized to create a news target with name " + newsTargetingEntity.getName(), currentIdentity.getUserId(), e);
```

File: services/src/main/java/org/exoplatform/news/service/NewsTargetingService.java [hunk 26]
```diff
@@ -101,4 +102,15 @@
    */
   void deleteNewsTargets(String newsId);
 
+  /**
+   * Create a news Metadata
+   * 
+   * @param newsTargetingEntity {@link News} TargetingEntity
+   * @param currentIdentity {@link Identity} technical identifier
+   * @return created {@link Metadata}
+   * @throws IllegalArgumentException when user create a metadata that already exist
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/NewsTargetingService.java:111]:
```suggestion
   * @throws IllegalArgumentException when user creates a {@link News} target that already exists
```

**Review by azayati — CHANGES_REQUESTED** [review]
see review

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 27]
```diff
@@ -202,4 +203,38 @@ public Response undoDeleteTarget(@Context HttpServletRequest request,
...
+      return Response.status(Response.Status.UNAUTHORIZED).entity(e.getMessage()).build();
+    } catch (IllegalArgumentException e) {
+      LOG.warn("User '{}' can't create a news targets '{}' with the same name " + newsTargetingEntity.getName(), currentIdentity.getUserId(), e);
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java:232]:
```suggestion
      LOG.warn("User '{}' can't create a news target with the same name " + newsTargetingEntity.getName(), currentIdentity.getUserId(), e);
```

**Review by azayati — CHANGES_REQUESTED** [review]
see review

**Review by azayati — DISMISSED** [review]
LGTM after QA validation

**Review by azayati — DISMISSED** [review]
LGTM

**Review by azayati — APPROVED** [review]
LGTM

**Pull request merged**